include/envoy/network/io_handle.h

```diff
@@ -7,27 +7,80 @@
 namespace Envoy {
 namespace Network {
 
+/**
+ * Base class for any I/O error.
+ */
+class IoError {
+public:
+  enum class IoErrorCode {
+    // Success.
+    NoError = 0,
+    // No data available right now, try again later.
+    Again,
+    // Not supported.
+    NoSupport,
+    // Address family not supported.
+    AddressFamilyNoSupport,
+    // During non-blocking connect, the connection cannot be completed immediately.
+    InProgress,
+    // Permission denied.
+    Permission,
+    // Other error codes cannot be mapped to any one above in getErrorCode().
+    UnknownError
+  };
+  virtual ~IoError() {}
+
+  // Map platform specific error into IoErrorCode.
+  virtual IoErrorCode getErrorCode() const PURE;
+
+  virtual std::string getErrorDetails() const PURE;
+};
+
+/**
+ * Basic type for return result which has a return code and error code defined
+ * according to different implementations.
+ */
+template <typename T> struct IoHandleCallResult {
+  IoHandleCallResult(T rc, std::unique_ptr<IoError> err) : rc_(rc), err_(std::move(err)) {}
+
+  IoHandleCallResult(IoHandleCallResult<T>&& result)
+      : rc_(result.rc_), err_(std::move(result.err_)) {}
+
+  virtual ~IoHandleCallResult() {}
+
+  T rc_;
+  std::unique_ptr<IoError> err_;
+};
+
+using IoHandleCallIntResult = IoHandleCallResult<int>;
```

Comment:
From a practical standpoint, can we just merge these? And when merging, since we now separate out the result from the error, can we avoid the negative and just have it return a uint64_t or something? Thoughts?

Reply:
Well, if success I would expect the error to be nullptr since it's a pointer reference, so can't the user tell? My preference is to merge them together and use an unsigned int for success, but curious what @alyssawilk thinks here.

Reply:
+1 to simplification if we can.

Rather than have one value for "did this work" and one for "what was the error" we could just have one enum for "what was the return value" where one code is success and the rest are error details.  

If we can avoid the pointers altogether it'd be good - doing a memory allocation per system call is going to be sad from a perf perspective.  As Matt points out that's hard to do with a virtual interface, but I'd almost prefer we skip the interface here and just make it a per-platform file, where things like QUIC and VPP fake the platform-specific error codes and we avoid a ton of news and frees.

Reply:
+1 on avoiding the allocation if we can, though I'm fine if you want to do it in a follow up and add a TODO. Either way.

Reply:
To address the allocation problem, instead of making IoError to have platform-dependent implementation, I propose to keep the interface and make err_ to point to a special address to indicate certain frequentlyl used error type, i.e. EAGAIN. So for all successful calls, err_ is nullptr, and for EAGAIN err_ points to a fixed address which is not actually newly allocated. Doing so allows us to have different implementations of IoError in same process. So that QuicIoHandle or VPPIoHandle can have their own definition of IoError if needed (though Quic probably won't).

This approach requires IoHandleCallResult::err_ to have a customized deleter to avoid deallocating memory not owned. So each implementation of IoHandle will need to provide such deleter.

I changed the code into this direction. PTAL!